### DISCUSSION TITLE
How to return custom error code if `onSubscribe` throws an error ?

### DISCUSSION THREAD

**Discussion by 97amarnathk** [enisdenjo/graphql-ws#167] [Q&A] [answered]
### Use Case
I am using graphql-ws with `authToken` which can expire. Incase the token expires when a client sends some graphql operation, I need to disconnect the socket and reconnect again.

### My approach
1. On each subscription event from the client, I am passing the  `authToken` in the `SubscriptionPayload`
2. `onSubscribe` validates the authToken , and if it has expired , it throws a `new Error("token invalid")`. Reference example: https://github.com/enisdenjo/graphql-ws#ws-auth-handling
3. In the client I am planning to add an `isFatalConnectionProblem` hook. Here I will see if the `error` has reason `"token invalid"`, get a new authToken, and retry again with this authToken.

### Problem
Because of an error code check [in client.ts](https://github.com/enisdenjo/graphql-ws/blob/master/src/client.ts#L470-L481),  returning the `"token invalid"` error  from `onSubscribe`, the websocket error code is 1011. Because of which `isFatalConnectionProblem` is not being called, and as a result the retry does not take place. 

**How do I throw a non 1011 error code to the client, if onSubscribe fails, so that I can retry connection?**

**Comment by enisdenjo** [comment, marked as answer]
Answered in full detail here: https://github.com/enisdenjo/graphql-ws/discussions/166#discussioncomment-653059.

TL;DR to close the socket manually with whatever close event you wish, you simply use the socket in the context. Something like `context.extra.socket.close(4401, 'Unauthorized')`.

**Reply by 97amarnathk** [reply]
So if I close the socket inside a graphql-ws server hook (onConnect or onSubscribe) to initate a new websocket connection, will there be any other side effects ?

Also:

if server sends back the `Unauthorized` request, I need to refresh my authToken inside `isFatalConnectionProblem`. However I am not able to do this because `isFatalConnectionProblem` is not async.

This can be done in the following ways:
* `connectionParams` method takes `errOrCloseEvent` as args, I can then use this to figure out a new set of `connectionParams`  (like refreshing authToken etc.)
* retryWait takes takes `errOrCloseEvent` and does the same as above

**Reply by enisdenjo** [reply]
Closing the socket at any point in the servers lifecycle should be safe, all cleanup and teardown will be done on close.

`isFatalConnectionProblem` should not be used for refreshing the token. It is not meant to be async because it is just a simple check whether the client should fail immediately or proceed with retrying.

As stated in the referenced comment above, you should be acquiring a new token in `connectionParams`, which is indeed async. You can check for validity there too and decide if you want to refresh or simply supply again the same token. Since it is async, youre free to take your time.

**Reply by enisdenjo** [reply]
If you cannot decide client side if the token expired (which you should), you can just have a flag that is something like `serverClosedBecauseUnauthorized` and set it to true once the close event matches the unauthorized close event dispatched by the server.

In that case, you can simply use the client's [`EventClosedListener`](https://github.com/enisdenjo/graphql-ws/blob/master/docs/modules/client.md#eventclosedlistener) and manipulate the flag from there.

**Reply by 97amarnathk** [reply]
Yes. 
We do store the expiry date in the token itself and just check this directly to avoid this call.

But an edge case is when the authToken is valid according to the client, but has been revoked/cancelled by the server (because of some other feature like force logout for users etc.)

So just a client side verification does not guarantee that the token is indeed valid.

**Reply by 97amarnathk** [reply]
>  you can just have a flag that is something like serverClosedBecauseUnauthorized and set it to true once the close event matches the unauthorized close event dispatched by the server.

How would I share this flag `serverClosedBecauseUnauthorized` between the 2 hooks `isFatalConnectionProblem` and `connectionParams` ?

**Reply by enisdenjo** [reply]
> But an edge case is when the authToken is valid according to the client, but has been revoked/cancelled by the server (because of some other feature like force logout for users etc.)

A valid point.

> How would I share this flag serverClosedBecauseUnauthorized between the 2 hooks isFatalConnectionProblem and connectionParams ?

Easy, just use JavaScript:
```js
import { createClient } from 'graphql-ws';
import { getCurrentToken, refreshCurrentToken } from './my-auth';

// indicates that the server closed the connection because of
// an auth problem. it indicates that the token should refresh
let shouldRefreshToken = false;

const client = createClient({
  url: 'ws://server-closed.because/unauthorized',
  connectionParams: async () => {
    if (shouldRefreshToken) {
      // refresh the token because it is no longer valid
      await refreshCurrentToken();
      // and reset the flag to avoid refreshing too many times
      shouldRefreshToken = false;
    }
    return { token: getCurrentToken() };
  },
  on: {
    closed: (event) => {
      // if closed with the `4401: Unauthorized` close event
      // server is communicating that the token is no longer
      // valid and should be therefore refreshed
      if (event.code === 4401) shouldRefreshToken = true;
    },
  },
});
```

**Reply by enisdenjo** [reply]
Hey @97amarnathk, I've added a [new recipe](https://github.com/enisdenjo/graphql-ws/commit/866889ff07da7d582b81a5e46400c3b20436640f) showcasing [ws server and client auth usage with token expiration, validation and refresh](https://github.com/enisdenjo/graphql-ws#auth-token). I sure hope it helps you get the concept described in our recent discussions.

**Reply by 97amarnathk** [reply]
Yes. Thanks a lot!
Works nicely. Only thing is 4401 too is a reserved code by graphql-ws, so is not retryable.

I am using 4430

**Reply by enisdenjo** [reply]
My bad! I changed the example to [use `4403: Forbidden` instead](https://github.com/enisdenjo/graphql-ws/commit/7642666be4e1e1e6d6394fbadb779f83a461cd7c).

Furthermore, I recommend using `4403` because [returning false from `onConnect` automatically closes the connection with a `4403: Forbidden`](https://github.com/enisdenjo/graphql-ws/blob/7642666be4e1e1e6d6394fbadb779f83a461cd7c/src/server.ts#L535-L537).
